src/docfx/build/Build.cs

```diff
@@ -14,46 +13,59 @@ internal static class Build
     {
         public static async Task Run(string docsetPath, CommandLineOptions options, ErrorLog errorLog)
         {
-            var repository = Repository.Create(docsetPath);
+            // load and trace entry repository
+            var repositoryProvider = new RepositoryProvider(docsetPath, options);
+            var repository = repositoryProvider.GetRepository(FileOrigin.Default);
             Telemetry.SetRepository(repository?.Remote, repository?.Branch);
+            var locale = LocalizationUtility.GetLocale(repository, options);
 
-            var locale = LocalizationUtility.GetLocale(repository?.Remote, repository?.Branch, options);
-            using (var restoreGitMap = GetRestoreGitMap(docsetPath, locale, options))
+            using (var restoreGitMap = RestoreGitMap.Create(docsetPath, locale))
             {
-                var fallbackRepo = GetFallbackRepository(docsetPath, repository, restoreGitMap);
+                // load configuration from current docset and fallback docset
+                var input = new Input(docsetPath, repositoryProvider);
+                var configLoader = new ConfigLoader(docsetPath, input, repositoryProvider);
+
+                repositoryProvider.ConfigRestoreMap(restoreGitMap);
+                var (errors, config) = configLoader.Load(options, extend: true);
 
-                var (configErrors, config) = GetBuildConfig(docsetPath, options, locale, fallbackRepo);
                 errorLog.Configure(config);
 
                 // just return if config loading has errors
-                if (errorLog.Write(configErrors))
+                if (errorLog.Write(errors))
                     return;
 
-                var (docset, fallbackDocset) = GetDocsetWithFallback(
-                    docsetPath, locale, config, repository, fallbackRepo, restoreGitMap);
-                var outputPath = Path.Combine(docsetPath, config.Output.Path);
-                var dependencyDocsets = LoadDependencies(docset, restoreGitMap);
+                // get docsets(build docset, fallback docset and dependency docsets)
+                repositoryProvider.Config(config);
+                var (docset, fallbackDocset) = GetDocsetWithFallback(docsetPath, locale, config, repositoryProvider, restoreGitMap);
+                if (!string.Equals(docset.DocsetPath, PathUtility.NormalizeFolder(docsetPath), PathUtility.PathComparison))
+                {
+                    // entry docset is not the docset to build
+                    input = new Input(docset.DocsetPath, repositoryProvider);
```

Comment:
Is there a way to detect if we are building from loc or building from English upfront?

Reply:
we don't until we get the configuration